### DISCUSSION TITLE
Don't show forgot password for Ldap user

### DISCUSSION THREAD

**Discussion by Bhavik237** [keycloak/keycloak#23042] [Q&A]
Hello everyone, I would like to hide the "Forgot Password" option on the login page for LDAP users. What changes should I make in the login FTL (FreeMarker Template Language) file to achieve this?

`  <div class="${properties.kcFormOptionsWrapperClass!}">
                            <#if realm.resetPasswordAllowed>
                                    <#if attributes?has_content && attributes.LDAP?has_content>
                                    <#else>
                                <span><a tabindex="5" href="${url.loginResetCredentialsUrl}">${msg("doForgotPassword")}</a></span>
                                    </#if>
                            </#if>
                        </div>`

if user is LDAP user he has attribute as LDAP 
i have added this , but it is not working in ftl file ,

Suggestion would be helpful
Thanks

**Comment by darius-m** [comment]
Are you sure the attributes field is (or should be) populated when you attempt to display the login form? What type of authentication page do you use - username + password, or just password after the username was asked in a different form? It feels like Keycloak should not attempt to fetch the user's information before confirming the password log in, but I assume there are cases where it may be required (e.g. for passwordless authentication).

**Reply by Bhavik237** [reply]
I am using username form then password form separately. okay I will look into passwordless authentication

**Comment by justin-tay** [comment]
You will likely need a custom `org.keycloak.forms.login.LoginFormsProviderFactory` that extends off `org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProviderFactory` to do this in addition to the FTL as the processing of the `LoginFormsPages.LOGIN_PASSWORD` page does not set attributes related to the user. For instance the code should resemble the code below.

```java
public class CustomLoginFormsProvider extends FreeMarkerLoginFormsProvider {

	public CustomLoginFormsProvider(KeycloakSession session) {
		super(session);
	}

	@Override
	protected Response createResponse(LoginFormsPages page) {
		if(LoginFormsPages.LOGIN_PASSWORD.equals(page)) {
			this.attributes.put("attributes", this.user.getAttributes());
		}
		return super.createResponse(page);
	}
}
```

**Reply by Bhavik237** [reply]
okay I will take a look at it.
